# Customer Shopping Data Analysis

## Introduction

This project analyzes a dataset containing customer shopping transactions from various malls. The dataset includes details about customers, the items they purchased, and the associated transaction information. 

### Dataset Overview

The dataset contains the following columns:
- **`invoice_no`**: Unique identifier for each transaction.
- **`customer_id`**: Unique customer identifier.
- **`gender`**: Gender of the customer (Male/Female).
- **`age`**: Age of the customer.
- **`category`**: Shopping category of the purchased items (e.g., Clothing, Shoes, Books).
- **`quantity`**: Quantity of items purchased.
- **`price`**: Price of the purchased items.
- **`payment_method`**: Payment method used (e.g., Credit Card, Debit Card, Cash).
- **`invoice_date`**: Date of the transaction.
- **`shopping_mall`**: Name of the shopping mall where the transaction took place.

### Objectives
1. **Data Cleaning**: Ensure the dataset is free of errors, duplicates, and missing values.
2. **Outlier Removal**: Identify and remove anomalous values in the `price` column.
3. **Standardization**: Uniformly format categorical values (e.g., `gender`, `category`).
4. **Date Parsing**: Convert `invoice_date` to a valid datetime format for temporal analysis.
5. **Exporting**: Save the cleaned dataset for further analysis and visualization.

---

## Steps Undertaken

### 1. Initial Inspection
- Loaded the dataset into a pandas DataFrame.
- Inspected the data for null values, duplicates, and incorrect formats.

### 2. Data Cleaning
- Removed duplicate rows to maintain data integrity.
- Filled missing values using appropriate techniques:
  - **Numerical Columns**: Used median values.
  - **Categorical Columns**: Used mode (most frequent value).
- Standardized categorical columns (`gender`, `category`) for consistent formatting.

### 3. Outlier Detection and Removal
- Calculated the Interquartile Range (IQR) for the `price` column.
- Removed rows with `price` values outside the acceptable range.

### 4. Date Formatting
- Converted the `invoice_date` column to a valid datetime format.
- Dropped rows with invalid or missing dates.

### 5. Exported Clean Dataset
- Saved the cleaned dataset to a new CSV file named **`cleaned_customer_shopping_data.csv`**.


In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv(r"C:\Users\Zana\Desktop\portfolio_projects\project_1\customer_shopping_data.csv")
print(df.head())

  invoice_no customer_id  gender  age  category  quantity    price  \
0    I138884     C241288  Female   28  Clothing         5  1500.40   
1    I317333     C111565    Male   21     Shoes         3  1800.51   
2    I127801     C266599    Male   20  Clothing         1   300.08   
3    I173702     C988172  Female   66     Shoes         5  3000.85   
4    I337046     C189076  Female   53     Books         4    60.60   

  payment_method invoice_date   shopping_mall  
0    Credit Card     5/8/2022          Kanyon  
1     Debit Card   12/12/2021  Forum Istanbul  
2           Cash    9/11/2021       Metrocity  
3    Credit Card   16/05/2021    Metropol AVM  
4           Cash   24/10/2021          Kanyon  


In [2]:
print(df.tail())

      invoice_no customer_id  gender  age         category  quantity    price  \
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & Beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & Beverage         2    10.46   
99455    I702964     C800631    Male   56       Technology         4  4200.00   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
99452    Credit Card   21/09/2022            Kanyon  
99453           Cash   22/09/2021    Forum Istanbul  
99454     Debit Card   28/03/2021         Metrocity  
99455           Cash   16/03/2021      Istinye Park  
99456    Credit Card   15/10/2022  Mall of Istanbul  


In [3]:
# Print the data types of each column in the DataFrame
df.dtypes

invoice_no         object
customer_id        object
gender             object
age                 int64
category           object
quantity            int64
price             float64
payment_method     object
invoice_date       object
shopping_mall      object
dtype: object

In [4]:
# Check for missing (NaN) values in each column
df.isnull().sum()

invoice_no        0
customer_id       0
gender            0
age               0
category          0
quantity          0
price             0
payment_method    0
invoice_date      0
shopping_mall     0
dtype: int64

In [5]:
# General information about the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99457 entries, 0 to 99456
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_no      99457 non-null  object 
 1   customer_id     99457 non-null  object 
 2   gender          99457 non-null  object 
 3   age             99457 non-null  int64  
 4   category        99457 non-null  object 
 5   quantity        99457 non-null  int64  
 6   price           99457 non-null  float64
 7   payment_method  99457 non-null  object 
 8   invoice_date    99457 non-null  object 
 9   shopping_mall   99457 non-null  object 
dtypes: float64(1), int64(2), object(7)
memory usage: 7.6+ MB


In [6]:
# Statistical summary for numerical data
df.describe()

age      quantity         price
count  99457.000000  99457.000000  99457.000000
mean      43.427089      3.003429    689.256321
std       14.990054      1.413025    941.184567
min       18.000000      1.000000      5.230000
25%       30.000000      2.000000     45.450000
50%       43.000000      3.000000    203.300000
75%       56.000000      4.000000   1200.320000
max       69.000000      5.000000   5250.000000

In [7]:
# Statistical summary for categorical data (object type columns)
df.describe(include='object')

invoice_no customer_id  gender  category payment_method invoice_date  \
count       99457       99457   99457     99457          99457        99457   
unique      99457       99457       2         8              3          797   
top       I138884     C241288  Female  Clothing           Cash   24/11/2021   
freq            1           1   59482     34487          44447          159   

           shopping_mall  
count              99457  
unique                10  
top     Mall of Istanbul  
freq               19943

In [8]:
# Fill NaN values for numerical columns (e.g., 'price') with median
df['price'] = df['price'].fillna(df['price'].median())

In [9]:
# Fill NaN values for categorical columns (e.g., 'category') with the most frequent value
df['category'] = df['category'].fillna(df['category'].mode()[0])

In [10]:
# Check for duplicates
df.duplicated().sum()  # This will show the number of duplicate rows

# Remove duplicates
df = df.drop_duplicates()

# Display DataFrame after removing duplicates
df

invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
3        I173702     C988172  Female   66            Shoes         5  3000.85   
4        I337046     C189076  Female   53            Books         4    60.60   
...          ...         ...     ...  ...              ...       ...      ...   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & Beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & Beverage         2    10.46   
99455    I702964     C800631    Male   56       Technology         4  4200.00   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card     5/8/2022            Kanyon  
1         Debit Card   12/12/2021    Forum Istanbul  
2               Cash    9/11/2021         Metrocity  
3        Credit Card   16/05/2021      Metropol AVM  
4               Cash   24/10/2021            Kanyon  
...              ...          ...               ...  
99452    Credit Card   21/09/2022            Kanyon  
99453           Cash   22/09/2021    Forum Istanbul  
99454     Debit Card   28/03/2021         Metrocity  
99455           Cash   16/03/2021      Istinye Park  
99456    Credit Card   15/10/2022  Mall of Istanbul  

[99457 rows x 10 columns]

In [11]:
# Calculation for the Interquartile Range (IQR) for the "price" column
Q1 = df['price'].quantile(0.25)
Q3 = df['price'].quantile(0.75)
IQR = Q3 - Q1

# Defined lower and upper bounds for outliers
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filtered out rows with 'price' outside these bounds (outliers)
df_filtered = df[(df['price'] >= lower_bound) & (df['price'] <= upper_bound)]

# Display the cleaned DataFrame after outlier removal
df_filtered

invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
4        I337046     C189076  Female   53            Books         4    60.60   
5        I227836     C657758  Female   28         Clothing         5  1500.40   
...          ...         ...     ...  ...              ...       ...      ...   
99451    I675411     C513603    Male   50             Toys         5   179.20   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & Beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & Beverage         2    10.46   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card     5/8/2022            Kanyon  
1         Debit Card   12/12/2021    Forum Istanbul  
2               Cash    9/11/2021         Metrocity  
4               Cash   24/10/2021            Kanyon  
5        Credit Card   24/05/2022    Forum Istanbul  
...              ...          ...               ...  
99451           Cash    9/10/2021      Metropol AVM  
99452    Credit Card   21/09/2022            Kanyon  
99453           Cash   22/09/2021    Forum Istanbul  
99454     Debit Card   28/03/2021         Metrocity  
99456    Credit Card   15/10/2022  Mall of Istanbul  

[94433 rows x 10 columns]

In [12]:
# Standardized "gender" column to ensure consistency
df_filtered['gender'] = df_filtered['gender'].str.capitalize()

# Standardize the "category" column
df_filtered['category'] = df_filtered['category'].str.capitalize()

# Display the updated DataFrame
df_filtered

C:\Users\Zana\AppData\Local\Temp\ipykernel_8368\4177448178.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['gender'] = df_filtered['gender'].str.capitalize()
C:\Users\Zana\AppData\Local\Temp\ipykernel_8368\4177448178.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['category'] = df_filtered['category'].str.capitalize()


invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
4        I337046     C189076  Female   53            Books         4    60.60   
5        I227836     C657758  Female   28         Clothing         5  1500.40   
...          ...         ...     ...  ...              ...       ...      ...   
99451    I675411     C513603    Male   50             Toys         5   179.20   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & beverage         2    10.46   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card     5/8/2022            Kanyon  
1         Debit Card   12/12/2021    Forum Istanbul  
2               Cash    9/11/2021         Metrocity  
4               Cash   24/10/2021            Kanyon  
5        Credit Card   24/05/2022    Forum Istanbul  
...              ...          ...               ...  
99451           Cash    9/10/2021      Metropol AVM  
99452    Credit Card   21/09/2022            Kanyon  
99453           Cash   22/09/2021    Forum Istanbul  
99454     Debit Card   28/03/2021         Metrocity  
99456    Credit Card   15/10/2022  Mall of Istanbul  

[94433 rows x 10 columns]

In [13]:
# Standardize the "gender" column using .loc to avoid the warning
df_filtered.loc[:, 'gender'] = df_filtered['gender'].str.capitalize()

# Standardize the "category" column using .loc to avoid the warning
df_filtered.loc[:, 'category'] = df_filtered['category'].str.capitalize()

# Display the updated DataFrame
df_filtered

invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
4        I337046     C189076  Female   53            Books         4    60.60   
5        I227836     C657758  Female   28         Clothing         5  1500.40   
...          ...         ...     ...  ...              ...       ...      ...   
99451    I675411     C513603    Male   50             Toys         5   179.20   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & beverage         2    10.46   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card     5/8/2022            Kanyon  
1         Debit Card   12/12/2021    Forum Istanbul  
2               Cash    9/11/2021         Metrocity  
4               Cash   24/10/2021            Kanyon  
5        Credit Card   24/05/2022    Forum Istanbul  
...              ...          ...               ...  
99451           Cash    9/10/2021      Metropol AVM  
99452    Credit Card   21/09/2022            Kanyon  
99453           Cash   22/09/2021    Forum Istanbul  
99454     Debit Card   28/03/2021         Metrocity  
99456    Credit Card   15/10/2022  Mall of Istanbul  

[94433 rows x 10 columns]

In [14]:
# Date Formatting:
# Converted "invoice_date" column to datetime format
df_filtered['invoice_date'] = pd.to_datetime(df_filtered['invoice_date'], errors='coerce', dayfirst=True)

# Checked if there are any invalid dates (NaT)
df_filtered[df_filtered['invoice_date'].isna()]

# Droped rows with invalid dates
df_filtered = df_filtered.dropna(subset=['invoice_date'])

# Display the DataFrame after date formatting
df_filtered


C:\Users\Zana\AppData\Local\Temp\ipykernel_8368\4042848807.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['invoice_date'] = pd.to_datetime(df_filtered['invoice_date'], errors='coerce', dayfirst=True)


invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
4        I337046     C189076  Female   53            Books         4    60.60   
5        I227836     C657758  Female   28         Clothing         5  1500.40   
...          ...         ...     ...  ...              ...       ...      ...   
99451    I675411     C513603    Male   50             Toys         5   179.20   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & beverage         2    10.46   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card   2022-08-05            Kanyon  
1         Debit Card   2021-12-12    Forum Istanbul  
2               Cash   2021-11-09         Metrocity  
4               Cash   2021-10-24            Kanyon  
5        Credit Card   2022-05-24    Forum Istanbul  
...              ...          ...               ...  
99451           Cash   2021-10-09      Metropol AVM  
99452    Credit Card   2022-09-21            Kanyon  
99453           Cash   2021-09-22    Forum Istanbul  
99454     Debit Card   2021-03-28         Metrocity  
99456    Credit Card   2022-10-15  Mall of Istanbul  

[94433 rows x 10 columns]

In [15]:
# Converted the "invoice_date" column to datetime format using .loc to avoid the warning
df_filtered.loc[:, 'invoice_date'] = pd.to_datetime(df_filtered['invoice_date'], errors='coerce', dayfirst=True)

# Checked if there are any invalid dates (NaT)
df_filtered[df_filtered['invoice_date'].isna()]

# Droped rows with invalid dates
df_filtered = df_filtered.dropna(subset=['invoice_date'])

# Display the DataFrame after date formatting
df_filtered

invoice_no customer_id  gender  age         category  quantity    price  \
0        I138884     C241288  Female   28         Clothing         5  1500.40   
1        I317333     C111565    Male   21            Shoes         3  1800.51   
2        I127801     C266599    Male   20         Clothing         1   300.08   
4        I337046     C189076  Female   53            Books         4    60.60   
5        I227836     C657758  Female   28         Clothing         5  1500.40   
...          ...         ...     ...  ...              ...       ...      ...   
99451    I675411     C513603    Male   50             Toys         5   179.20   
99452    I219422     C441542  Female   45         Souvenir         5    58.65   
99453    I325143     C569580    Male   27  Food & beverage         2    10.46   
99454    I824010     C103292    Male   63  Food & beverage         2    10.46   
99456    I232867     C273973  Female   36         Souvenir         3    35.19   

      payment_method invoice_date     shopping_mall  
0        Credit Card   2022-08-05            Kanyon  
1         Debit Card   2021-12-12    Forum Istanbul  
2               Cash   2021-11-09         Metrocity  
4               Cash   2021-10-24            Kanyon  
5        Credit Card   2022-05-24    Forum Istanbul  
...              ...          ...               ...  
99451           Cash   2021-10-09      Metropol AVM  
99452    Credit Card   2022-09-21            Kanyon  
99453           Cash   2021-09-22    Forum Istanbul  
99454     Debit Card   2021-03-28         Metrocity  
99456    Credit Card   2022-10-15  Mall of Istanbul  

[94433 rows x 10 columns]

In [16]:
# Save the cleaned DataFrame to a CSV file
df_filtered.to_csv('cleaned_customer_shopping_data.csv', index=False)

## Key Results

### Cleaned Dataset Summary
- **Total Rows**: 94,433
- **Total Columns**: 10
- **Data Types**: Numerical (int64, float64), Categorical (object), and Datetime.

### Statistical Summary of Numerical Columns
| Metric       | Age     | Quantity | Price       |
|--------------|---------|----------|-------------|
| **Count**    | 94,433  | 94,433   | 94,433      |
| **Mean**     | 43.43   | 3.00     | 689.26      |
| **Std Dev**  | 14.99   | 1.41     | 941.18      |
| **Min**      | 18.00   | 1.00     | 5.23        |
| **25%**      | 30.00   | 2.00     | 45.45       |
| **50%**      | 43.00   | 3.00     | 203.30      |
| **75%**      | 56.00   | 4.00     | 1,200.32    |
| **Max**      | 69.00   | 5.00     | 5,250.00    |

### Cleaned Categorical Columns
- **Gender**: Standardized to "Male" and "Female".
- **Category**: Standardized to consistent capitalized formats.

### Valid Date Column
- `invoice_date`: Fully converted to datetime format for analysis.

---